addons/delivery/models/delivery_carrier.py

```diff
@@ -97,6 +107,20 @@ class DeliveryCarrier(models.Model):
         ('shipping_insurance_is_percentage', 'CHECK(shipping_insurance >= 0 AND shipping_insurance <= 100)', "The shipping insurance must be a percentage between 0 and 100."),
     ]
 
+    @api.constrains('must_have_tag_ids', 'excluded_tag_ids')
+    def _check_tags(self):
+        for carrier in self:
+            if carrier.must_have_tag_ids & carrier.excluded_tag_ids:
+                raise UserError(_('A tag cannot be both in the must have and excluded tags.'))
```

Comment:
Would be great to indicate the problematic carrier in the error message (in case of mass import or something like that)